pkg/util/watch/until.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watch
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	watchapi "k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	watchtools "k8s.io/client-go/tools/watch"
+	"k8s.io/klog/v2"
+	"kpt.dev/configsync/pkg/kinds"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// UntilDeletedWithSync watches the specified object and blocks until it is
+// fully deleted (NotFound, not just deleteTimestamp). This allows for any
+// finalizers to complete.
+//
+// If the object is not yet deleted, the delete call will use foreground
+// deletion, waiting until garbage collection completes too. However, if the
+// object was already deleted
+//
+// If the object is Unstructured, the GVK must be set.
+//
+// The ResourceVersion of the input object is ignored, because a new Informer is
+// created which will do a ListAndWatch, and List needs the collection
+// ResourceVersion instead of the object ResourceVersion.
```

Comment:
This is technically incorrect. There is no such thing as an object RV. There is only the RV of the whole resource. We could in fact just do a Watch with "time-travel", if we already have the RV of the object we want to see deleted.